pkg/main/server.go

```diff
@@ -0,0 +1,138 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+)
+
+// Server an interface for managing function servers
+type Server interface {
+	IsIdle() bool
+	SetIdle(bool)
+	GetPort() uint16
+	Invoke(input map[string]interface{}) (io.ReadCloser, error)
+	Stdout() *bytes.Buffer
+	Stderr() *bytes.Buffer
+	Start() error
+	Shutdown() error
+}
+
+// DefaultServer a struct to hold information about running servers
+type DefaultServer struct {
+	isIdle bool
+	port   uint16
+	cmd    *exec.Cmd
+	client *http.Client
+	stdout *bytes.Buffer
+	stderr *bytes.Buffer
+}
+
+// IsIdle indicates whether this server is currently idle or processing a request
+func (s *DefaultServer) IsIdle() bool {
+	return s.isIdle
+}
+
+// SetIdle sets whether this server is currently idle or processing a request
+func (s *DefaultServer) SetIdle(idle bool) {
+	s.isIdle = idle
+}
+
+// GetPort returns the port this server is running on
+func (s *DefaultServer) GetPort() uint16 {
+	return s.port
+}
+
+// Invoke calls the server with the given input to invoke a Dispatch function
+func (s *DefaultServer) Invoke(input map[string]interface{}) (io.ReadCloser, error) {
+	p, err := json.Marshal(input)
+
+	ctx, ok := input["context"].(map[string]interface{})
+	if !ok {
+		return nil, BadRequestError(p)
+	}
+	timeout := 0.0
+	if ctx["timeout"] != nil {
+		timeout = ctx["timeout"].(float64)
+	}
+
+	s.client.Timeout = time.Duration(timeout) * time.Millisecond
+
+	resp, err := s.client.Post(fmt.Sprintf("http://0.0.0.0:%d", s.GetPort()), "application/json", bytes.NewBuffer(p))
+
+	if err != nil {
+		println(err.Error())
+		if strings.Contains(err.Error(), "Client.Timeout") {
+			return nil, TimeoutError(fmt.Sprintf("%d", timeout))
+		} else if strings.Contains(err.Error(), "connection refused") {
+			return nil, errors.New("connection refused")
+		}
+	}
+
+	if resp.StatusCode == 400 {
+		return nil, BadRequestError("invalid input")
+	} else if resp.StatusCode == 500 {
+		return nil, InvocationError(resp.StatusCode)
+	} else if resp.StatusCode == 422 {
+		return nil, InvalidResponsePayloadError("")
+	} else if resp.StatusCode == 502 {
+		return nil, UnknownSystemError("")
+	}
+
+	return resp.Body, nil
+}
+
+// Stdout returns the Buffer containing stdout
+func (s *DefaultServer) Stdout() *bytes.Buffer {
+	return s.stdout
+}
+
+// Stderr returns the Buffer containing stderr
+func (s *DefaultServer) Stderr() *bytes.Buffer {
+	return s.stderr
+}
+
+// Start starts the server
+func (s *DefaultServer) Start() error {
+	return s.cmd.Start()
+}
+
+// Shutdown shuts down the server, kills it if necessary
+func (s *DefaultServer) Shutdown() error {
+	err := s.cmd.Wait()
+	if err != nil {
+		return s.cmd.Process.Kill()
+	}
+
+	return err
+}
+
+// NewServer returns a new DefaultServer with the given port and command
+func NewServer(port uint16, cmd *exec.Cmd) (*DefaultServer, error) {
```

Comment:
NIT: constructors are usually placed near the type definition, somewhere near the top of the file.